### PR TITLE
Automate parameter function creation of BatchGrad and Grad extractors

### PR DESCRIPTION
**Pull request by tjs111**
To be more flexible with parameter names, refactor the firstorder extension such that arbitrary parameter names are supported.
Adding a support for a new module would then not require any additional changes to the firstorder modules for batch_grad.
Please review whether the docstrings are sufficient to understand the structure and purpose.
Please review whether the other firstorder modules should also be refactored.

### CODE REVIEW

File: backpack/extensions/firstorder/batch_grad/batch_grad_base.py [hunk 1]
```diff
@@ -1,17 +1,40 @@
...
+
+        If the param method has already been defined, it is left unchanged.
+
+        Args:
+            derivatives(backpack.core.derivatives.basederivatives.BaseParameterDerivatives):# noqa
+                Derivatives object assigned to self.derivatives.
+            params (list[str]): list of strings with parameter names. Defaults to None.
+        """
         self.derivatives = derivatives
+        for param_str in params:
```
Comment (f-dangel) [inline backpack/extensions/firstorder/batch_grad/batch_grad_base.py:29]: Per default, `params` is `None`, which will raise an exception because `None` is not iterable. There are two solutions:

- Remove `params`' default value
- Only loop if `params` is not `None`

What's the better choice?

File: backpack/extensions/firstorder/batch_grad/batch_grad_base.py [hunk 2]
```diff
@@ -1,17 +1,40 @@
...
+            if not hasattr(self, param_str):
+                self.__setattr__(param_str, self._make_function(param_str))
         super().__init__(params=params)
 
-    def bias(self, ext, module, g_inp, g_out, bpQuantities):
-        return self.derivatives.bias_jac_t_mat_prod(
-            module, g_inp, g_out, g_out[0], sum_batch=False
-        )
+    def _make_function(self, param):
+        def function(ext, module, g_inp, g_out, bpQuantities):
```
Comment (f-dangel) [inline backpack/extensions/firstorder/batch_grad/batch_grad_base.py:35]: * Make function names more descriptive
* Add documentation

File: backpack/extensions/firstorder/batch_grad/batch_grad_base.py [hunk 3]
```diff
@@ -1,17 +1,40 @@
...
     def __init__(self, derivatives, params=None):
+        """Initializes all methods.
+
+        If the param method has already been defined, it is left unchanged.
+
+        Args:
+            derivatives(backpack.core.derivatives.basederivatives.BaseParameterDerivatives):# noqa
+                Derivatives object assigned to self.derivatives.
+            params (list[str]): list of strings with parameter names. Defaults to None.
+        """
```
Comment (f-dangel) [inline backpack/extensions/firstorder/batch_grad/batch_grad_base.py:27]: The documentation is quite general. I think it would be easier to understand if it was referring more specifically to the `BatchGrad` extension.

File: backpack/extensions/firstorder/batch_grad/batch_grad_base.py [hunk 4]
```diff
@@ -1,17 +1,40 @@
...
 class BatchGradBase(FirstOrderModuleExtension):
+    """Passes the calls for the parameters to the derivatives class.
+
+    This class implements functions with method names from params.
+
+    If child class wants to overwrite these methods
+    - for example to support an additional external module -
+    it can do so using the interface for parameter "param1"
+    param1(ext, module, g_inp, g_out, bpQuantities):
+        return batch_grads
```
Comment (f-dangel) [inline backpack/extensions/firstorder/batch_grad/batch_grad_base.py:14]: Apply code-style formatting

File: backpack/extensions/firstorder/gradient/base.py [hunk 5]
```diff
@@ -1,17 +1,40 @@
...
+        def function(ext, module, g_inp, g_out, bpQuantities):
+            return getattr(self.derivatives, f"{param}_jac_t_mat_prod")(
+                module, g_inp, g_out, g_out[0], sum_batch=True
+            )
 
-    def weight(self, ext, module, g_inp, g_out, bpQuantities):
-        return self.derivatives.weight_jac_t_mat_prod(
-            module, g_inp, g_out, g_out[0], sum_batch=True
-        )
+        return function
```
Comment (f-dangel) [inline backpack/extensions/firstorder/gradient/base.py:40]: Same feedback as for `firstorder/batch_grad`

File: backpack/extensions/firstorder/batch_grad/batch_grad_base.py (hunk 1 above)
Comment (tjs111) [reply backpack/extensions/firstorder/batch_grad/batch_grad_base.py:29]: Removing the default value is the better choice in my opinion.
A third alternative would be to change the default value to an empty list. But I think removing it altogether is better.

File: backpack/extensions/firstorder/batch_grad/batch_grad_base.py (hunk 3 above)
Comment (tjs111) [reply backpack/extensions/firstorder/batch_grad/batch_grad_base.py]: I changed the documentation such that it says which quantities are calculated. If you mean something different by "specifically to the BatchGrad extension", please explain it to me.

File: backpack/extensions/firstorder/batch_grad/batch_grad_base.py [hunk 6]
```diff
@@ -15,26 +16,48 @@ class BatchGradBase(FirstOrderModuleExtension):
...
                 Derivatives object assigned to self.derivatives.
-            params (list[str]): list of strings with parameter names. Defaults to None.
+            params (list[str]): list of strings with parameter names.
+                For each, a method is assigned.
         """
         self.derivatives = derivatives
         for param_str in params:
             if not hasattr(self, param_str):
-                self.__setattr__(param_str, self._make_function(param_str))
+                self.__setattr__(param_str, self._make_param_function(param_str))
```
Comment (f-dangel) [inline backpack/extensions/firstorder/batch_grad/batch_grad_base.py:33]: more a personal choice to avoid dunder methods, but does `setattr(self, param_str, self._make_param_function(param_str))` work, too?

File: backpack/extensions/firstorder/batch_grad/batch_grad_base.py [hunk 7]
```diff
@@ -15,26 +16,48 @@ class BatchGradBase(FirstOrderModuleExtension):
...
+
+            Args:
+                ext(backpack.extensions.BatchGrad): extension that is used
+                module(torch.nn.Module): module that performed forward pass
+                g_inp(tuple[torch.Tensor]): input gradient tensors
+                g_out(tuple[torch.Tensor]): output gradient tensors
+                bpQuantities(None): additional quantities for second order
+
+            Returns:
+                torch.Tensor: individual gradients
```
Comment (f-dangel) [inline backpack/extensions/firstorder/batch_grad/batch_grad_base.py:57]: scaled individual gradients, as there might be a factor `1 / batch_size`, depending on the loss reduction

File: backpack/extensions/firstorder/gradient/base.py [hunk 8]
```diff
@@ -1,12 +1,15 @@
-"""Passes the calls for the parameters to the derivatives class."""
+"""Calculates the gradient."""
 from backpack.extensions.firstorder.base import FirstOrderModuleExtension
 
 
 class GradBaseModule(FirstOrderModuleExtension):
-    """Passes the calls for the parameters to the derivatives class.
+    """calculates the gradient.
```
Comment (f-dangel) [inline backpack/extensions/firstorder/gradient/base.py:6]: Capitalize

File: backpack/extensions/firstorder/batch_grad/batch_grad_base.py (hunk 4 above)
Comment (tjs111) [reply backpack/extensions/firstorder/batch_grad/batch_grad_base.py]: I added a line before and after. I didn't find out how to add multiline code.

File: backpack/extensions/firstorder/batch_grad/batch_grad_base.py (hunk 3 above)
Comment (f-dangel) [reply backpack/extensions/firstorder/batch_grad/batch_grad_base.py]: Maybe something along [these lines](https://gist.github.com/redlotus/3bc387c2591e3e908c9b63b97b11d24e#file-docstrings-py-L270-L274) works?

File: backpack/extensions/firstorder/batch_grad/batch_grad_base.py (hunk 4 above)
Comment (f-dangel) [reply backpack/extensions/firstorder/batch_grad/batch_grad_base.py]: Maybe something along [these lines](https://gist.github.com/redlotus/3bc387c2591e3e908c9b63b97b11d24e#file-docstrings-py-L270-L274) works?

File: backpack/extensions/firstorder/gradient/base.py [hunk 9]
```diff
@@ -1,17 +1,67 @@
...
 
 
 class GradBaseModule(FirstOrderModuleExtension):
-    def __init__(self, derivatives, params=None):
+    """Calculates the gradient.
+
+    This class implements functions with method names from params.
+
+    Passes the calls for the parameters to the derivatives class.
+    Implements functions with method names from params.
```
Comment (f-dangel) [inline backpack/extensions/firstorder/gradient/base.py:11]: First and last line seem redundant. Remove duplicate?

File: backpack/extensions/firstorder/gradient/base.py [hunk 10]
```diff
@@ -1,17 +1,67 @@
...
+    In this case, the method is not overwritten by this class.
+    """
+
+    def __init__(self, derivatives, params):
+        """Initializes all methods.
+
+        If the param method has already been defined, it is left unchanged.
+
+        Args:
+            derivatives(backpack.core.derivatives.basederivatives.BaseParameterDerivatives):# noqa
```
Comment (f-dangel) [inline backpack/extensions/firstorder/gradient/base.py:29]: Can you specify the exact error code that should be ignored here by the docstring checker? [Here](https://github.com/f-dangel/cockpit/blob/9ea1dd4fe60d43f4cec0220f6e5f19e4ae81ed9e/cockpit/quantities/hooks/base.py#L25) is an example. Otherwise, other potential errors will be ignored, too.

File: backpack/extensions/firstorder/batch_grad/batch_grad_base.py [hunk 11]
```diff
@@ -1,17 +1,65 @@
...
+    In this case, the method is not overwritten by this class.
+    """
+
+    def __init__(self, derivatives, params):
+        """Initializes all methods.
+
+        If the param method has already been defined, it is left unchanged.
+
+        Args:
+            derivatives(backpack.core.derivatives.basederivatives.BaseParameterDerivatives):# noqa
```
Comment (f-dangel) [inline backpack/extensions/firstorder/batch_grad/batch_grad_base.py:27]: Can you specify the exact error code that should be ignored here by the docstring checker? [Here](https://github.com/f-dangel/cockpit/blob/9ea1dd4fe60d43f4cec0220f6e5f19e4ae81ed9e/cockpit/quantities/hooks/base.py#L25) is an example. Otherwise, other potential errors will be ignored, too.
